File: calendar1/src/Service/CalendarServiceTokenRing.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.ServiceModel;
using System.Threading;
using Microsoft.Samples.XmlRpc;

namespace BitCalendarService
{
    public class CalendarServiceTokenRing : ICalendarService
    {
        private CalendarDatabaseManager _calendarDatabaseManager    = new CalendarDatabaseManager();
        private static  Thread          _tokenThread                = new Thread(TokenCoroutine);
        private const   string          _baseAddress                = "/BitCalendarService";

        private static  bool            _requestingCriticalRegion   = false;
        private static  bool            _accessingCriticalRegion    = false;
        private static  bool            _hasToken                   = false;

        static CalendarServiceTokenRing()
        {
            
        }
        public CalendarServiceTokenRing()
        {
            
        }

        public static void TokenCoroutine()
        {
            while (true)
            {
                if (_hasToken)
                {
                    while (_requestingCriticalRegion || _accessingCriticalRegion)
                    {
                        Console.WriteLine("Passing Thread: Using token!");
                    }

                    Console.WriteLine("Passing thread: Token free! Passing.");

                    // Setup a channel for remote service.
                    _hasToken = false;
                    var remoteChannelFactory    = CreateChannelFactory(CalendarServiceUtility.IPAndPort);
                    var remoteCalendarService   = remoteChannelFactory.CreateChannel();
                    remoteCalendarService.PassToken();
                    remoteChannelFactory.Close();

                    Console.WriteLine("Passing thread: Token passing complete.");
                }
            }
        }
        public bool     Initialize()
        {
            if (!_tokenThread.IsAlive)
            {
                Console.WriteLine("Starting token thread.");
                _tokenThread.Start();
            }

            if (!DecodeUsers(GetUsers()).Any(x => x.IPAddress == CalendarServiceUtility.IPAndPort) && DecodeUsers(GetUsers()).Count() == 0)
                RegisterUser(CalendarServiceUtility.IPAndPort);
            
            Console.WriteLine("Initialization complete.");

            return true;
        }
        public bool     PassToken()
        {
            // We are done with token. Attempt passing it.
            _hasToken = false;

            // Get index of the next user.
            var users       = DecodeUsers(GetUsers()).ToList();
            var ownIndex    = users.IndexOf(users.Find(x => x.IPAddress == CalendarServiceUtility.IPAndPort));
            var tarIndex    = ownIndex + 1 == users.Count ? 0 : ownIndex + 1;

            Console.WriteLine("Initial target index is: " + tarIndex);

            // Pass it to the next available user.
            var result = false;
            while (!result)
            {
                try
                {
                    // Setup a channel for remote service.
                    var remoteChannelFactory    = CreateChannelFactory(users[tarIndex].IPAddress);
                    var remoteCalendarService   = remoteChannelFactory.CreateChannel();

                    Console.WriteLine("Passing token to: " + users[tarIndex].IPAddress);

                    // Attempt to make it receieve token.
                    remoteCalendarService.ReceiveToken();

                    Console.WriteLine("Passed token to: " + users[tarIndex].IPAddress);

                    result = true;
                }
                catch (Exception)
                {
                    result   = false;
                    tarIndex = tarIndex + 1 == users.Count ? 0 : tarIndex + 1;
                }
            }

            return result;
        }
        public bool     ReceiveToken()
        {
            Console.WriteLine("Received token!");

            _hasToken = true;

            return true;
        }

        public bool     CreateCalendarEvent(string calendarEventString)
        {
            _requestingCriticalRegion = true;

            // If we are connected and we don't have token, wait.
            if(DecodeUsers(GetUsers()).Length > 1)
                while (!_hasToken) ;

            _accessingCriticalRegion = true;

            var result = true; // CreateCalendarEventInternal(calendarEventString);

            var users = DecodeUsers(GetUsers());
            foreach (var user in users)
            {
                var channelFactory  = CreateChannelFactory(user.IPAddress);
                var calendarService = channelFactory.CreateChannel();
                if (!calendarService.CreateCalendarEventInternal(calendarEventString))
                    result = false;
                channelFactory.Close();
            }

            _requestingCriticalRegion   = false;
            _accessingCriticalRegion    = false;

            return result;
        }
        public bool     EditCalendarEvent(string calendarEventString)
        {
            _requestingCriticalRegion = true;

            // If we are connected and while we don't have token.
            if (DecodeUsers(GetUsers()).Length > 1)
                while (!_hasToken) ;

            _accessingCriticalRegion = true;

            var result = true; // EditCalendarEventInternal(calendarEventString);

            var users = DecodeUsers(GetUsers());
            foreach (var user in users)
            {
                var channelFactory  = CreateChannelFactory(user.IPAddress);
                var calendarService = channelFactory.CreateChannel();
                if (!calendarService.EditCalendarEventInternal(calendarEventString))
                {
                    Console.WriteLine("Edit failed. Attempting add.");
                    if (!calendarService.CreateCalendarEventInternal(calendarEventString))
                        result = false;
                }
                channelFactory.Close();
            }

            _requestingCriticalRegion   = false;
            _accessingCriticalRegion    = false;

            return result;
        }
        public bool     DeleteCalendarEvent(string calendarEventString)
        {
            _requestingCriticalRegion = true;

            // If we are connected and while we don't have token.
            if (DecodeUsers(GetUsers()).Length > 1)
                while (!_hasToken) ;

            _accessingCriticalRegion = true;

            var result = true; // DeleteCalendarEventInternal(calendarEventString);

            var users = DecodeUsers(GetUsers());
            foreach (var user in users)
            {
                var channelFactory = CreateChannelFactory(user.IPAddress);
                var calendarService = channelFactory.CreateChannel();
                if (!calendarService.DeleteCalendarEventInternal(calendarEventString))
                result = false;
                channelFactory.Close();
            }

            _requestingCriticalRegion   = false;
            _accessingCriticalRegion    = false;

            return result;
        }
        public string   GetCalendarEvents()
        {
            var calendarEventsString = "";

            var calendarEvents = _calendarDatabaseManager.GetCalendarEvents();
            var calendarEventStrings = new string[calendarEvents.Length];
            for (var i = 0; i < calendarEvents.Length; i++)
            {
                calendarEventStrings[i] = calendarEvents[i].EncodeToString();
                calendarEventsString += calendarEventStrings[i];
                if (i + 1 < calendarEvents.Length)
                    calendarEventsString += "---";
            }

            return calendarEventsString;
        }
        public string   GetUsers()
        {
            var calendarUsersString = "";

            var calendarUsers = _calendarDatabaseManager.GetUsers();
            var calendarUserStrings = new string[calendarUsers.Length];
            for (var i = 0; i < calendarUsers.Length; i++)
            {
                calendarUserStrings[i] = calendarUsers[i].EncodeToString();
                calendarUsersString += calendarUserStrings[i];
                if (i + 1 < calendarUsers.Length)
                    calendarUsersString += "---";
            }
            return calendarUsersString;
        }

        public bool     Connect(string targetIPAddress)
        {
            bool result;

            Console.WriteLine("Attempting to connect to IP: " + targetIPAddress);

            try
            {
                // Setup a channel for remote service.
                var remoteChannelFactory    = CreateChannelFactory(targetIPAddress);
                var remoteCalendarService   = remoteChannelFactory.CreateChannel();

                Console.WriteLine("Retrieving users list.");

                // Sync users from the target PC (add yourself too for token ring).
                var usersString = remoteCalendarService.GetUsers() + "---" + CalendarServiceUtility.IPAndPort;

                Console.WriteLine("Retrieved users list.");
                Console.WriteLine("Syncing users.");

                result = SyncUsers(usersString);

                Console.WriteLine("Synced users.");
                Console.WriteLine("Syncing calendar events.");

                if(result)
                    result = SyncCalendarEvents(remoteCalendarService.GetCalendarEvents());

                Console.WriteLine("Synced calendar events.");

                remoteChannelFactory.Close();

                // Register to other people's user table.
                var users = DecodeUsers(GetUsers());
                foreach (var user in users)
                {
                    if (user.IPAddress == CalendarServiceUtility.IPAndPort)
                        continue;

                    var othersChannelFactory    = CreateChannelFactory(user.IPAddress);
                    var othersCalendarService   = othersChannelFactory.CreateChannel();
                    if (!othersCalendarService.RegisterUser(CalendarServiceUtility.IPAndPort))
                        result = false;
                    othersChannelFactory.Close();
                }

                // If network consists of me and someone else (if network was just initiated):
                if (users.Length == 2) /* make this two */
                {
                    ReceiveToken();
                }
            }
            catch (Exception)
            {
                Console.WriteLine("Exception connecting to network from IP: " + targetIPAddress);
                return false;
            }

            return result;
        }
        public bool     Disconnect()
        {
            // You cannot disconnect until you pass the token.
            while (_hasToken) ;

            var result = true;

            var users = DecodeUsers(GetUsers());
            foreach (var user in users)
            {
                var othersChannelFactory = CreateChannelFactory(user.IPAddress);
                var othersCalendarService = othersChannelFactory.CreateChannel();
                if (!othersCalendarService.DeregisterUser(CalendarServiceUtility.IPAndPort))
                    result = false;
                othersChannelFactory.Close();
            }

            ClearUsersTable();

            return result;
        }
        public bool     RegisterUser(string calendarUserIPString)
        {
            if (_calendarDatabaseManager.CreateUser(new CalendarUser(calendarUserIPString)))
                return true;
            return false;
        }
        public bool     DeregisterUser(string calendarUserIPString)
        {
            if (_calendarDatabaseManager.DeleteUser(new CalendarUser(calendarUserIPString)))
                return true;
            return false;
        }

        public bool     ClearCalendarEventsTable()
        {
            _calendarDatabaseManager.ClearCalendarEvents();
            return true;
        }
        public bool     ClearUsersTable()
        {
            _calendarDatabaseManager.ClearUsers();

            if (!DecodeUsers(GetUsers()).Any(x => x.IPAddress == CalendarServiceUtility.IPAndPort) && DecodeUsers(GetUsers()).Count() == 0)
                RegisterUser(CalendarServiceUtility.IPAndPort);

            return true;
        }

        public bool     CreateCalendarEventInternal(string calendarEventString)
        {
            var result = _calendarDatabaseManager.CreateCalendarEvent(new CalendarEvent(calendarEventString));
            return result;
        }
        public bool     EditCalendarEventInternal(string calendarEventString)
        {
            var result = _calendarDatabaseManager.EditCalendarEvent(new CalendarEvent(calendarEventString));
            return result;
        }
        public bool     DeleteCalendarEventInternal(string calendarEventString)
        {
            var result = _calendarDatabaseManager.DeleteCalendarEvent(new CalendarEvent(calendarEventString));
            return result;
        }

        public string   RequestLock(string timestampString)
        {
            throw new NotImplementedException();
        }
        public bool     DebugLock()
        {
            _requestingCriticalRegion = true;

            return true;
        }
        public bool     DebugUnlock()
        {
            _requestingCriticalRegion = false;

            return true;
        }

        private bool    SyncUsers(string calendarUsersString)
        {
            if (calendarUsersString == null)
                return true;

            string[] stringSeperators = new string[] { "---" };
            string[] splitList = calendarUsersString.Split(stringSeperators, StringSplitOptions.None);

            var calendarUsers = new List<CalendarUser>();
            for (var i = 0; i < splitList.Length; i++)
            {
                var calendarUser = new CalendarUser();
                if (calendarUser.DecodeFromString(splitList[i]))
                    calendarUsers.Add(calendarUser);
            }

            _calendarDatabaseManager.ClearUsers();
            _calendarDatabaseManager.CreateUsers(calendarUsers.ToArray());

            return true;
        }
        private bool    SyncCalendarEvents(string calendarEventsString)
        {
            if (calendarEventsString == null)
                return true;
            string[] stringSeperators = new string[] { "---" };
            string[] splitList = calendarEventsString.Split(stringSeperators, StringSplitOptions.None);

            var calendarEvents = new List<CalendarEvent>();
            for (var i = 0; i < splitList.Length; i++)
            {
                var calendarEvent = new CalendarEvent();
                if (calendarEvent.DecodeFromString(splitList[i]))
                    calendarEvents.Add(calendarEvent);
            }

            // Should we really delete all previous events when connecting?
            _calendarDatabaseManager.ClearCalendarEvents();
            _calendarDatabaseManager.CreateCalendarEvents(calendarEvents.ToArray());

            return true;
        }

        private CalendarUser[]                              DecodeUsers(string calendarUsersString)
        {
            if (calendarUsersString == null)
                return new CalendarUser[0];

            string[] stringSeperators = new string[] { "---" };
            string[] splitList = calendarUsersString.Split(stringSeperators, StringSplitOptions.None);

            var calendarUsers = new List<CalendarUser>();
            for (int i = 0; i < splitList.Length; i++)
            {
                var calendarUser = new CalendarUser();
                if (calendarUser.DecodeFromString(splitList[i]))
                    calendarUsers.Add(calendarUser);
            }

            return calendarUsers.ToArray();
        }
        private CalendarEvent[]                             DecodeCalendarEvents(string calendarEventsString)
        {
            if (calendarEventsString == null)
                return new CalendarEvent[0];

            string[] stringSeperators = new string[] { "---" };
            string[] splitList = calendarEventsString.Split(stringSeperators, StringSplitOptions.None);

            var calendarEvents = new List<CalendarEvent>();
            for (int i = 0; i < splitList.Length; i++)
            {
                var calendarEvent = new CalendarEvent();
                if (!string.IsNullOrEmpty(splitList[i]) && calendarEvent.DecodeFromString(splitList[i]))
                    calendarEvents.Add(calendarEvent);
            }

            return calendarEvents.ToArray();
        }
        private static ChannelFactory<ICalendarService>     CreateChannelFactory(string ipAndPort)
        {
            var address = new Uri("http://" + ipAndPort + Program.BaseAddress);
            var channelFactory =
                new ChannelFactory<ICalendarService>(new WebHttpBinding(WebHttpSecurityMode.None),
                                                     new EndpointAddress(address));
            channelFactory.Endpoint.Behaviors.Add(new XmlRpcEndpointBehavior());

            Console.WriteLine("Created channel factory to: " + address);

            return channelFactory;
        }
    }
}

File: calendar2/src/Service/CalendarServiceTokenRing.java

import java.net.URL;
import java.util.ArrayList;
import org.apache.xmlrpc.client.XmlRpcClient;
import org.apache.xmlrpc.client.XmlRpcClientConfigImpl;

public class CalendarServiceTokenRing implements ICalendarService {

    private CalendarDatabaseManager _calendarDatabaseManager = new CalendarDatabaseManager();
    private static Thread _tokenThread;

    private static volatile boolean _requestingCriticalRegion = false;
    private static volatile boolean _accessingCriticalRegion = false;
    private static volatile boolean _hasToken = false;

    public boolean CreateCalendarEvent(String calendarEventString) {
        _requestingCriticalRegion = true;

        // If we are connected and we don't have token, wait.
        if (DecodeUsers(GetUsers()).length > 1) {
            while (!_hasToken) {

            }
        }

        // Start accessing critical region.
        _accessingCriticalRegion = true;

        boolean result = true;

        CalendarUser[] users = DecodeUsers(GetUsers());
        for (CalendarUser user : users) {
            XmlRpcClient remoteServiceChannel = CreateServiceChannel(user.IPAddress);
            try {
                Object[] params = new Object[]{calendarEventString};
                boolean newResult = (boolean) remoteServiceChannel.execute("BitCalendarService.CreateCalendarEventInternal", params);
                if (!newResult) {
                    result = false;
                }
            } catch (Exception e) {
                System.out.println("Remote create error!");
            }

        }

        _requestingCriticalRegion = false;
        _accessingCriticalRegion = false;

        return result;
    }

    public boolean EditCalendarEvent(String calendarEventString) {
        _requestingCriticalRegion = true;

        if (DecodeUsers(GetUsers()).length > 1) {
            while (!_hasToken) {

            }
        }
        // Start accessing critical region.
        _accessingCriticalRegion = true;

        boolean result = true;

        CalendarUser[] users = DecodeUsers(GetUsers());
        for (CalendarUser user : users) {
            XmlRpcClient remoteServiceChannel = CreateServiceChannel(user.IPAddress);
            try {
                Object[] params = new Object[]{calendarEventString};
                boolean newResult = (boolean) remoteServiceChannel.execute("BitCalendarService.EditCalendarEventInternal", params);
                if (!newResult) {
                    System.out.println("Edit failed. Attempting add.");
                    newResult = (boolean) remoteServiceChannel.execute("BitCalendarService.AddCalendarEventInternal", params);
                    if (!newResult) {
                        result = false;
                    }
                }
            } catch (Exception e) {
                System.out.println("Remote edit error!");
            }

        }

        _requestingCriticalRegion = false;
        _accessingCriticalRegion = false;

        return result;
    }

    public boolean DeleteCalendarEvent(String calendarEventString) {
        _requestingCriticalRegion = true;
        if (DecodeUsers(GetUsers()).length > 1) {
            while (!_hasToken) {

            }
        }
        // Start accessing critical region.
        _accessingCriticalRegion = true;

        boolean result = true;

        CalendarUser[] users = DecodeUsers(GetUsers());
        for (CalendarUser user : users) {
            XmlRpcClient remoteServiceChannel = CreateServiceChannel(user.IPAddress);
            try {
                Object[] params = new Object[]{calendarEventString};
                boolean newResult = (boolean) remoteServiceChannel.execute("BitCalendarService.DeleteCalendarEventInternal", params);
                if (!newResult) {
                    result = false;
                }
            } catch (Exception e) {
                System.out.println("Remote delete error!");
            }

        }

        _requestingCriticalRegion = false;
        _accessingCriticalRegion = false;

        return result;
    }

    public String GetCalendarEvents() {
        String calendarEventsString = "";

        _calendarDatabaseManager = new CalendarDatabaseManager();

        CalendarEvent[] calendarEvents = _calendarDatabaseManager.GetCalendarEvents();
        String[] calendarEventStrings = new String[calendarEvents.length];
        for (int i = 0; i < calendarEvents.length; i++) {
            calendarEventStrings[i] = calendarEvents[i].EncodeToString();
            calendarEventsString += calendarEventStrings[i];
            if (i + 1 < calendarEvents.length) {
                calendarEventsString += "---";
            }
        }

        return calendarEventsString;
    }

    public boolean RegisterUser(String calendarUserString) {
        _calendarDatabaseManager = new CalendarDatabaseManager();
        return _calendarDatabaseManager.CreateUser(new CalendarUser(calendarUserString));
    }

    public boolean DeregisterUser(String calendarUserString) {
        _calendarDatabaseManager = new CalendarDatabaseManager();
        return _calendarDatabaseManager.DeleteUser(new CalendarUser(calendarUserString));
    }

    public String GetUsers() {
        String calendarUsersString = "";

        _calendarDatabaseManager = new CalendarDatabaseManager();

        CalendarUser[] calendarUsers = _calendarDatabaseManager.GetUsers();
        String[] calendarUserStrings = new String[calendarUsers.length];
        for (int i = 0; i < calendarUsers.length; i++) {
            calendarUserStrings[i] = calendarUsers[i].EncodeToString();
            calendarUsersString += calendarUserStrings[i];
            if (i + 1 < calendarUsers.length) {
                calendarUsersString += "---";
            }
        }

        return calendarUsersString;
    }

    public boolean SyncUsers(String calendarUsersString) {
        if (calendarUsersString == null) {
            return true;
        }

        String[] splitArray = calendarUsersString.split("\\-\\-\\-");

        ArrayList<CalendarUser> calendarUsers = new ArrayList<>();
        for (int i = 0; i < splitArray.length; i++) {
            CalendarUser newUser = new CalendarUser();
            if (newUser.DecodeFromString(splitArray[i])) {
                calendarUsers.add(newUser);
            }
        }

        _calendarDatabaseManager = new CalendarDatabaseManager();

        _calendarDatabaseManager.ClearUsers();
        _calendarDatabaseManager.CreateUsers(calendarUsers.toArray(new CalendarUser[0]));

        return true;
    }

    public boolean SyncCalendarEvents(String calendarEventsString) {
        if (calendarEventsString == null) {
            return true;
        }

        String[] splitArray = calendarEventsString.split("\\-\\-\\-");

        ArrayList<CalendarEvent> calendarEvents = new ArrayList<>();
        for (int i = 0; i < splitArray.length; i++) {
            CalendarEvent newEvent = new CalendarEvent();
            if (newEvent.DecodeFromString(splitArray[i])) {
                calendarEvents.add(newEvent);
            }
        }

        _calendarDatabaseManager = new CalendarDatabaseManager();

        _calendarDatabaseManager.ClearCalendarEvents();
        _calendarDatabaseManager.CreateCalendarEvents(calendarEvents.toArray(new CalendarEvent[0]));

        return true;
    }

    public boolean ClearCalendarEventsTable() {
        _calendarDatabaseManager = new CalendarDatabaseManager();
        _calendarDatabaseManager.ClearCalendarEvents();
        return true;
    }

    public boolean ClearUsersTable() {
        _calendarDatabaseManager = new CalendarDatabaseManager();
        _calendarDatabaseManager.ClearUsers();

        boolean localExist = false;
        CalendarUser[] users = DecodeUsers(GetUsers());
        for (CalendarUser user : users) {
            if (user.IPAddress == Server.IPAndPort()) {
                localExist = true;
            }
        }
        if (!localExist && users.length == 0) {
            RegisterUser(Server.IPAndPort());
        }

        return true;
    }

    public static void TokenCoroutine() {
        while (true) {
            if (_hasToken) {
                while (_requestingCriticalRegion || _accessingCriticalRegion) {
                    System.out.println("Passing Thread: Using token!");
                }

                System.out.println("Passing thread: Token free! Passing.");

                XmlRpcClient remoteServiceChannel = CreateServiceChannel(Server.IPAndPort());
                try {
                    // Setup a channel for remote service.
                    _hasToken = false;
                    Object[] params = new Object[]{};
                    boolean newResult = (boolean) remoteServiceChannel.execute("BitCalendarService.PassToken", params);

                } catch (Exception e) {
                    System.out.println("Passing thread: Token passing failed! Trying next user!");
                }

                System.out.println("Passing thread: Token passing complete.");
            }
        }
    }

    public boolean Connect(String targetIPAndPort) {
        boolean result;

        System.out.println("Attempting to connect to IP: " + targetIPAndPort);

        try {
            XmlRpcClient remoteServiceChannel = CreateServiceChannel(targetIPAndPort);
            Object[] params;

            System.out.println("Retrieving users list.");

            // Sync users from the target PC.
            params = new Object[]{};
            String usersString = (String) remoteServiceChannel.execute("BitCalendarService.GetUsers", params);
            usersString += "---" + Server.IPAndPort();

            System.out.println("Retrieved users list.");
            System.out.println("Syncing users.");

            result = SyncUsers(usersString);

            System.out.println("Synced users.");
            System.out.println("Retrieving calendar events list.");

            params = new Object[]{};
            String calendarEventsString = (String) remoteServiceChannel.execute("BitCalendarService.GetCalendarEvents", params);

            System.out.println("Retrieved calendar events list.");
            System.out.println("Syncing calendar events.");

            if (result) {
                result = SyncCalendarEvents(calendarEventsString);
            }

            System.out.println("Synced calendar events.");

            // Register to other people's user table.
            CalendarUser[] users = DecodeUsers(GetUsers());
            for (CalendarUser user : users) {

                if (user.IPAddress.equals(Server.IPAndPort()))
                        continue;
                
                remoteServiceChannel = CreateServiceChannel(user.IPAddress);

                params = new Object[]{Server.IPAndPort()};
                boolean newResult = (boolean) remoteServiceChannel.execute("BitCalendarService.RegisterUser", params);

                if (!newResult) {
                    result = false;
                }
            }

            // If network consists of me and someone else (if network was just initiated):
            if (users.length == 2) /* make this two */ {
                ReceiveToken();
            }

        } catch (Exception e) {
            System.out.println("Exception connecting to network from IP: " + targetIPAndPort);
            return false;
        }

        return result;
    }

    public boolean Disconnect() {
        boolean result = true;

        CalendarUser[] users = DecodeUsers(GetUsers());
        for (CalendarUser user : users) {
            XmlRpcClient remoteServiceChannel = CreateServiceChannel(user.IPAddress);
            try {
                Object[] params = new Object[]{Server.IPAndPort()};
                boolean newResult = (boolean) remoteServiceChannel.execute("BitCalendarService.DeregisterUser", params);
                if (!newResult) {
                    result = false;
                }
            } catch (Exception e) {
                System.out.println("Deregister error!");
            }
        }

        ClearUsersTable();

        return result;
    }

    public boolean CreateCalendarEventInternal(String calendarEventString) {
        _calendarDatabaseManager = new CalendarDatabaseManager();
        return _calendarDatabaseManager.CreateCalendarEvent(new CalendarEvent(calendarEventString));
    }

    public boolean EditCalendarEventInternal(String calendarEventString) {
        _calendarDatabaseManager = new CalendarDatabaseManager();
        return _calendarDatabaseManager.EditCalendarEvent(new CalendarEvent(calendarEventString));

    }

    public boolean DeleteCalendarEventInternal(String calendarEventString) {
        _calendarDatabaseManager = new CalendarDatabaseManager();
        return _calendarDatabaseManager.DeleteCalendarEvent(new CalendarEvent(calendarEventString));

    }

    public String RequestLock(String timestampString) {
        return "";
    }

    public boolean Initialize() {   
        System.out.println("Initializing.");
        
        //if (_tokenThread == null || !_tokenThread.isAlive()) {
            _tokenThread = new Thread(new TokenThread());
            _tokenThread.start();
        //}

        boolean localExist = false;
        CalendarUser[] users = DecodeUsers(GetUsers());
        for (CalendarUser user : users) {
            if (user.IPAddress == Server.IPAndPort()) {
                localExist = true;
            }
        }
        if (!localExist && users.length == 0) {
            RegisterUser(Server.IPAndPort());
        }

        System.out.println("Initialization complete.");

        return true;
    }

    public boolean PassToken() {
        boolean result = false;
 
        // We are done with token. Attempt passing it.
        _hasToken = false;

        int ownIndex = -1;
        int tarIndex;
        
        // Get index of the next user.
        CalendarUser[] users = DecodeUsers(GetUsers());
        for(int i = 0; i < users.length; i++)
        {
            if(users[i].IPAddress.equals(Server.IPAndPort()))
            {
                System.out.println("Found own index: " + ownIndex);
                ownIndex = i;
                break;
            }
        }
        tarIndex = ownIndex + 1 == users.length ? 0 : ownIndex + 1;

        System.out.println("Initial target index is: " + tarIndex);

        // Pass it to the next available user.
        while (!result) 
        {
            try 
            {
                    System.out.println("Passing token to: " + users[tarIndex].IPAddress);

                    XmlRpcClient remoteServiceChannel = CreateServiceChannel(users[tarIndex].IPAddress);

                    Object[] params = new Object[]{};
                    boolean newResult = (boolean) remoteServiceChannel.execute("BitCalendarService.ReceiveToken", params);
                    
                    System.out.println("Passed token to: " + users[tarIndex].IPAddress);
                    
                    result = true;
            }
            catch(Exception e)
            {
                System.out.println("Pass exception!");
                result   = false;
                tarIndex = tarIndex + 1 == users.length ? 0 : tarIndex + 1;
            }
        }
        return result;
    }

    public boolean ReceiveToken() {
        System.out.println("Received token!");

        _hasToken = true;

        return true;
    }

    public boolean DebugLock() {
        _requestingCriticalRegion = true;

        return true;
    }

    public boolean DebugUnlock() {
        _requestingCriticalRegion = false;

        return true;
    }

    private CalendarUser[] DecodeUsers(String calendarUsersString) {
        String[] splitArray = calendarUsersString.split("\\-\\-\\-");

        ArrayList<CalendarUser> calendarUsers = new ArrayList<>();
        for (int i = 0; i < splitArray.length; i++) {
            CalendarUser newUser = new CalendarUser();
            if (newUser.DecodeFromString(splitArray[i])) {
                calendarUsers.add(newUser);
            }
        }

        return calendarUsers.toArray(new CalendarUser[0]);
    }

    private CalendarEvent[] DecodeCalendarEvents(String calendarEventsString) {
        if (calendarEventsString == null) {
            return new CalendarEvent[0];
        }

        String[] splitArray = calendarEventsString.split("\\-\\-\\-");

        ArrayList<CalendarEvent> calendarEvents = new ArrayList<>();
        for (int i = 0; i < splitArray.length; i++) {
            CalendarEvent newEvent = new CalendarEvent();
            if (newEvent.DecodeFromString(splitArray[i])) {
                calendarEvents.add(newEvent);
            }
        }

        return calendarEvents.toArray(new CalendarEvent[0]);
    }

    private static XmlRpcClient CreateServiceChannel(String ipAndPort) {
        System.out.println("Creating service channel to: http://" + ipAndPort + Server.BaseAddress);
        XmlRpcClientConfigImpl config = new XmlRpcClientConfigImpl();
        try {
            config.setServerURL(new URL("http://" + ipAndPort + Server.BaseAddress));
        } catch (Exception e) {
            e.printStackTrace();
        }

        XmlRpcClient client = new XmlRpcClient();
        client.setConfig(config);

        return client;
    }
}
